newIDE/app/src/AssetStore/CreditsPackages/CreditsPackagePurchaseDialog.js

```diff
@@ -0,0 +1,346 @@
+  const [isPurchasing, setIsPurchasing] = React.useState(false);
+  const [purchaseSuccessful, setPurchaseSuccessful] = React.useState(false);
+  const [
+    displayPasswordPrompt,
+    setDisplayPasswordPrompt,
+  ] = React.useState<boolean>(false);
+  const [password, setPassword] = React.useState<string>('');
+  const { showAlert } = useAlertDialog();
+
+  const shouldUseOrSimulateAppStoreProduct =
+    shouldUseAppStoreProduct() || simulateAppStoreProduct;
+
+  const onStartPurchase = async () => {
+    if (!profile) return;
+    setDisplayPasswordPrompt(false);
+
+    // Purchase with the App Store.
+    if (shouldUseOrSimulateAppStoreProduct) {
+      try {
+        setIsPurchasing(true);
+        await purchaseAppStoreProduct(
+          creditsPackageListingData.appStoreProductId
+        );
+      } finally {
+        setIsPurchasing(false);
+      }
+      return;
+    }
+
+    // Purchase with web.
+    try {
+      setIsPurchasing(true);
+      const checkoutUrl = await getPurchaseCheckoutUrl({
+        productId: creditsPackageListingData.id,
+        priceName: creditsPackageListingData.prices[0].name,
+        userId: profile.id,
+        userEmail: profile.email,
+        password,
+      });
+      Window.openExternalURL(checkoutUrl);
+    } catch (error) {
+      const extractedStatusAndCode = extractGDevelopApiErrorStatusAndCode(
+        error
+      );
+      if (
+        extractedStatusAndCode &&
+        extractedStatusAndCode.status === 403 &&
+        extractedStatusAndCode.code === 'auth/wrong-password'
+      ) {
+        await showAlert({
+          title: t`Operation not allowed`,
+          message: t`The password you entered is incorrect. Please try again.`,
+        });
+      } else {
+        console.error('Unable to get the checkout URL', error);
+        await showAlert({
+          title: t`An error happened`,
+          message: t`Unable to get the checkout URL. Please try again later.`,
+        });
+      }
+      setIsPurchasing(false);
+    } finally {
+      setPassword('');
+    }
+  };
+
+  const onWillPurchase = () => {
+    // Password is required in dev environment only so that one cannot freely purchase game templates.
+    if (Window.isDev()) setDisplayPasswordPrompt(true);
+    else onStartPurchase();
+  };
+
+  React.useEffect(
+    () => {
+      onWillPurchase();
+    },
+    // Launch the start process directly when the dialog is opened, to avoid an extra click.
+    // eslint-disable-next-line
+    []
+  );
+
+  useInterval(
+    () => {
+      onRefreshLimits();
+    },
+    isPurchasing ? 3900 : null
+  );
+
+  // Listen to the limits being loaded the first time (either when the dialog is opened, or when the user logs in)
+  // In this case, save the user balance before the purchase.
+  React.useEffect(
+    () => {
+      (async () => {
+        if (limits && initialCreditsAmount === null) {
+          setInitialCreditsAmount(limits.credits.userBalance.amount);
```

Comment:
I feel like this variable should be a ref since you don't want anything to change when the value changes. But I guess it works this way

Reply:
Yeah I could test with a ref, and yet it works like that so will probably not touch it 😅